src/delayedarray/Subset.py

```diff
@@ -0,0 +1,302 @@
+from typing import Tuple, Sequence
+from collections import namedtuple
+from copy import deepcopy
+from numpy import dtype, array, ndarray, ix_
+from bisect import bisect_left
+
+from .interface import extract_dense_array, extract_sparse_array, is_sparse
+from .SparseNdarray import SparseNdarray
+from .utils import sanitize_single_index
+
+
+# Check if the subset is already unique and/or sorted. If it is, we can optimize
+# some of the downstream procedures.
+def _is_unique_and_sorted(subset):
+    is_sorted = True
+    for i in range(1, len(subset)):
+        if subset[i] < subset[i-1]:
+            is_sorted = False
+            break
+
+    is_unique = True
+    if is_sorted:
+        for i in range(1, len(subset)):
+            if subset[i] == subset[i-1]:
+                is_unique = False
+                break
+    else:
+        occurred = set()
+        for s in subset:
+            if s in occurred:
+                is_unique = False
+                break
+            occurred.add(s)
```

Comment:
could be simplified to

`is_unique = len(subset) != len(set(subset))` ?

Reply:
For brevity. Maybe less so for performance, as these loops can short-circuit if one duplicate is found.